### PR TITLE
Remove exceptions from the strip unpacker

### PR DESCRIPTION
**Pull request by pieterdavid**
#### PR description:

These changes address https://github.com/cms-sw/cmssw/issues/24825 (also tracked in jira as [CMSTRACK-161](https://its.cern.ch/jira/browse/CMSTRACK-161)): the strip unpacker does not throw exceptions anymore (those left in EventFilter/SiStripRawToDigi are only used in the packer, and signal a bad input, so I think they are acceptable). Instead, status codes are returned.
The extra information that used to be in the exception message is printed through `LogDebug`, so it won't be there in release builds anymore, but the warnings (by default: maximally once per type, and then a count at the end of the job) that were printed are still there (minus the information that came from the exception message - FED and channel ID are still there, these are the most relevant anyway).

Most potential exceptions came from the `sistrip::FEDBuffer` (and `sistrip::FEDSpyBuffer`, together with their common base class `sistrip::FEDBufferBase`) constructor, so I moved the checks this does into a separate method that should be called before constructing a FED(Spy)Buffer, and returns a status code. For the FEDBuffer class there is also an initialisation step that could throw; I made it public and return a status code as well; a typical FEDBuffer construction then looks like this:
```C++
const auto st_buffer = sistrip::preconstructCheckFEDBuffer(input);
// check st_buffer
sistrip::FEDBuffer buffer{input};
const auto st_chan = buffer.findChannels();
// check st_chan
```
see [the AlignmentEventFilter diff](https://github.com/cms-sw/cmssw/compare/master...pieterdavid:stripunpacker_exceptions?expand=1#diff-65baa272872b1b76219e23f67163e9b5) for a full example (behaviour should be the same as before for this, except for the exception message) - this is also mentioned in the methods' documentation now.

Since I had to change the `operator++()` functions of the channel unpackers (from where some exceptions could be thrown), and these were all used from a loop like this:
```C++
while ( unpacker.hasData() ) {
  output.push_back(...); // (raw) digi with unpacker adc (and strip, for ZS)
  unpacker++;
}
```
I changed them into methods (see [here](https://github.com/pieterdavid/cmssw/blob/stripunpacker_exceptions/EventFilter/SiStripRawToDigi/interface/SiStripFEDBuffer.h#L181-L324)) that take an output iterator instead, and unpack the whole channel (this allows for a bit of simplification, and makes the code more readable; this was already all inlined before, so I don't expect significant changes in CPU performance, but in principle a few more things could be optimised by the compiler now)<sup>1</sup>.
With this, I could also move the code that select the correct unpacking mode depending on the readout mode and packet code into a helper method (the earlier changes in https://github.com/cms-sw/cmssw/pull/24339 to the regional unpacker already went in this direction), that is shared between the "full" and "regional" unpacker (as a side-effect, the latter now supports all modes, except scope mode, which is only used for calibration, as far as I know), and quite a bit of code could be removed from these plugins.

These changes are purely technical, no changes are expected (also not on the computing performance; most of the unpacker time should be spent in allocating memory for all the digis, and the way that is done is not changed by this PR).

#### PR validation:

I took 100 events from a file taken in virgin raw mode (2015 HI, with 10bits packing), unpacked and saved all the raw digis, with the unmodified release. Then I ran, for all VR modes, a packer+unpacker combination on top of those and compared the raw digis before and after; likewise for the ZS modes, after running the zero-suppression - no unexpected differences ([config](https://github.com/pieterdavid/SiStripLocalRecoDebugging/blob/master/test/unpackVRAndCompare.py)).
`runTheMatrix.py -l 140.55` (2018 HI hybrid emulation + repacking + reco workflow, this is one of the more exotic unpacker use cases)
If anyone knows of data with problems that should trigger one of these exceptions-turned-warnings, that would be very useful to test as well.

CC: @robervalwalsh @mmusich @tsusa @alesaggio

<hr/> 

<sup>1</sup> in passing I made a few small changes to SiStripRawDigi (which is nothing more than a `uint16_t` that inherits from `edm::DoNotSortUponInsertion`) - the one that was convenient is conversion back to `uint16_t`, but I think the other ones are sensible and binary compatible (I don't know of anywhere these are persisted anyway).

### CODE REVIEW

File: DataFormats/SiStripDigi/interface/SiStripRawDigi.h [hunk 1]
```diff
@@ -14,12 +14,14 @@
...
-  SiStripRawDigi(const uint16_t& adc) : adc_(adc) { ; }
+  explicit SiStripRawDigi(uint16_t adc) : adc_(adc) { ; }
 
   SiStripRawDigi() : adc_(0) { ; }
   ~SiStripRawDigi() { ; }
 
-  inline const uint16_t& adc() const;
+  inline uint16_t adc() const { return adc_; }
+
+  inline operator uint16_t() const { return adc_; }
```
Comment (slava77) [inline DataFormats/SiStripDigi/interface/SiStripRawDigi.h]: I guess that this should at least be explicit, if present at all.

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by jfernan2** [comment]
please test

**Comment by jfernan2** [comment]
+1

**Comment by civanch** [comment]
+1

File: EventFilter/SiStripRawToDigi/plugins/SiStripRawToDigiUnpacker.cc [hunk 2]
```diff
@@ -1048,7 +815,7 @@ namespace sistrip {
```
Comment (slava77) [inline EventFilter/SiStripRawToDigi/plugins/SiStripRawToDigiUnpacker.cc]: static analyzer is complaining about const cast away on L812
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-5e6c19/4284/llvm-analysis/report-cd0929.html#EndPath
It looks like ```const TFHeaderDescription* header = (const TFHeaderDescription*)data_u32;``` or an appropriate cast will fix the warning.

I have this and a few other const_cast cleanups in this package in 
slava77:CMSSW_11_1_0_pre1/test28756/pull28756-31d9c24 https://github.com/cms-sw/cmssw/commit/98b95af12c91c77f2bcce2e271acfe8dba452d48

Comment (pieterdavid) [reply EventFilter/SiStripRawToDigi/plugins/SiStripRawToDigiUnpacker.cc]: Thanks! I cherry-picked the commit, but implementing the const getter with the non-const version resulted in an infinite recursion, so I went back to doing it the other way around https://github.com/cms-sw/cmssw/pull/28756/commits/f0c4e61934314a36230ed2394b440374fcd726de (using `std::as_const`, as suggested in https://stackoverflow.com/a/47369227/11079521)

**Comment by christopheralanwest** [comment]
please test

File: EventFilter/SiStripRawToDigi/plugins/SiStripRawToDigiUnpacker.cc (hunk 2 above)
Comment (slava77) [reply EventFilter/SiStripRawToDigi/plugins/SiStripRawToDigiUnpacker.cc]: I only checked that it compiles. Thank you for checking the full details.
I suspect that the static analyzer will still complain here. I guess we can (later?) add a note "CMS_SA_ALLOW" in front of the implementation line.
@makortel @Dr15Jones do you have a better suggestion?

**Comment by jfernan2** [comment]
+1

File: EventFilter/SiStripRawToDigi/plugins/SiStripRawToDigiUnpacker.cc (hunk 2 above)
Comment (slava77) [reply EventFilter/SiStripRawToDigi/plugins/SiStripRawToDigiUnpacker.cc]: so, the static analyzer indeed does not like it; but based on https://stackoverflow.com/a/47369227/11079521, it looks like the pattern is coming from recommended literature.
I consider this an acceptable solution then.

**Comment by slava77** [comment]
+1

for https://github.com/cms-sw/cmssw/pull/28756	f0c4e61 
- code changes are in line with the PR description and the follow up review. 
- jenkins tests pass and comparisons with the baseline show no differences, as expected from this mainly technical change

**Comment by civanch** [comment]
+1

**Comment by slava77** [comment]
@christopheralanwest @tocheng   we need an alca signature for this PR.
Please check and comment if there are issues.
Thank you.

File: EventFilter/SiStripRawToDigi/plugins/SiStripRawToDigiUnpacker.cc (hunk 2 above)
Comment (makortel) [reply EventFilter/SiStripRawToDigi/plugins/SiStripRawToDigiUnpacker.cc]: > but based on https://stackoverflow.com/a/47369227/11079521, it looks like the pattern is coming from recommended literature.
> I consider this an acceptable solution then.

I agree.

The static analyzer warning can indeed be silenced once(/if) we get the static analyzer to support applying `CMS_SA_ALLOW` to `const_cast`.

**Comment by silviodonato** [comment]
This is just a friendly reminder to @christopheralanwest @tocheng

**Comment by christopheralanwest** [comment]
+1

**Comment by silviodonato** [comment]
+1

**Pull request merged**